src/v/pandaproxy/server.h

```diff
@@ -208,8 +208,17 @@ class auth_ctx_server : public ctx_server<service_t> {
             case config::rest_authn_method::http_basic: {
                 return dispatch([this, func{std::forward<Func>(func)}](
                                   kafka_client_cache& cache) mutable {
-                    auto client = cache.fetch_or_insert(user, authn_method);
-                    return std::invoke(std::forward<Func>(func), *client)
+                    auto item = cache.fetch_or_insert(user, authn_method);
+                    client_ptr client = item.first;
+                    client_sm_ptr client_sm = item.second;
+                    return ss::with_semaphore(
+                             *client_sm,
+                             1,
+                             [client,
+                              func{std::forward<Func>(func)}]() mutable {
+                                 return std::invoke(
+                                   std::forward<Func>(func), *client);
+                             })
```

Comment:
Maybe it's worth encapsulating this?

```diff
diff --git a/src/v/pandaproxy/server.h b/src/v/pandaproxy/server.h
index a5ae182f1..eb775243d 100644
--- a/src/v/pandaproxy/server.h
:...skipping...
diff --git a/src/v/pandaproxy/kafka_client_cache.h b/src/v/pandaproxy/kafka_client_cache.h
index bb9cba309..5fb30287d 100644
--- a/src/v/pandaproxy/kafka_client_cache.h
+++ b/src/v/pandaproxy/kafka_client_cache.h
@@ -57,6 +57,17 @@ public:
     std::pair<client_ptr, client_sm_ptr>
     fetch_or_insert(credential_t user, config::rest_authn_method authn_method);
 
+    template<
+      std::invocable<kafka::client::client&> Func,
+      typename Futurator
+      = ss::futurize<std::invoke_result_t<Func, kafka::client::client&>>>
+    typename Futurator::type with_client_for(
+      credential_t user, config::rest_authn_method authn_method, Func&& func) {
+        auto [c, s] = fetch_or_insert(std::move(user), authn_method);
+        auto units = co_await ss::get_units(*s, 1);
+        co_return co_await Futurator::invoke(std::forward<Func>(func), *c);
+    }
+
     ss::future<> clean_stale_clients();
 
     size_t size() const;
diff --git a/src/v/pandaproxy/server.h b/src/v/pandaproxy/server.h
index a5ae182f1..eb775243d 100644
--- a/src/v/pandaproxy/server.h
+++ b/src/v/pandaproxy/server.h
@@ -214,18 +214,8 @@ public:
             case config::rest_authn_method::http_basic: {
                 return dispatch([this, func{std::forward<Func>(func)}](
                                   kafka_client_cache& cache) mutable {
-                    auto item = cache.fetch_or_insert(user, authn_method);
-                    client_ptr client = item.first;
-                    client_sm_ptr client_sm = item.second;
-                    return ss::with_semaphore(
-                             *client_sm,
-                             1,
-                             [client,
-                              func{std::forward<Func>(func)}]() mutable {
-                                 return std::invoke(
-                                   std::forward<Func>(func), *client);
-                             })
-                      .finally([client] {});
+                    return cache.with_client_for(
+                      user, authn_method, std::forward<Func>(func));
                 });
             }
             }
@@ -241,18 +231,8 @@ public:
             case config::rest_authn_method::http_basic: {
                 return dispatch([this, func{std::forward<Func>(func)}](
                                   kafka_client_cache& cache) mutable {
-                    auto item = cache.fetch_or_insert(user, authn_method);
-                    client_ptr client = item.first;
-                    client_sm_ptr client_sm = item.second;
-                    return ss::with_semaphore(
-                             *client_sm,
-                             1,
-                             [client,
-                              func{std::forward<Func>(func)}]() mutable {
-                                 return std::invoke(
-                                   std::forward<Func>(func), *client);
-                             })
-                      .finally([client] {});
+                    return cache.with_client_for(
+                      user, authn_method, std::forward<Func>(func));
                 });
             }
             }
```

Reply:
So bring back `invoke_on_cache` :)

Reply:
This is nit and can also be handled later.

Reply:
Done.